### PR TITLE
Add canine in documentation_tests_file

### LINKED ISSUE #16292: Doc tests sprint
### This issue is part of our **Doc Test Sprint**. If you're interested in helping out come [join us on Discord](https://discord.gg/J8bW9u5abB) and talk with other contributors!

Docstring examples are often the first point of contact when trying out a new library! So far we haven't done a very good job at ensuring that all docstring examples work correctly in 🤗 Transformers - but we're now very dedicated to ensure that all documentation examples work correctly by integrating Python's doctest (https://docs.python.org/3/library/doctest.html) into our daily github action tests. 

In short we should add doc tests for all of our models for both PyTorch and Tensorflow. So far we've only covered ca. 30 PyTorch models as can be seen here: https://github.com/huggingface/transformers/blob/master/utils/documentation_tests.txt . 

Adding a documentation test for a model is a great way to better understand how the model works, a simple (possibly first) contribution to Transformers and most importantly a very important contribution to the Transformers community 🔥 

If you're interested in adding a documentation test, please read through the **Guide to contributing** below.

This issue is a call for contributors, to make sure docstring exmaples of existing model architectures work correctly. If you wish to contribute, reply in this thread which architectures you'd like to take :)

### Guide to contributing:
1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your architecture(s) in this thread (confirm no one is working on it) 🎯 
3. Implement the changes as in https://github.com/huggingface/transformers/pull/15987 (see the diff on the model architectures for a few examples) 💪 
    - The file you want to look at is in `src/transformers/models/[model_name]/modeling_[model_name].py`, `src/transformers/models/[model_name]/modeling_tf_[model_name].py` or `src/transformers/doc_utils.py` or `src/transformes/file_utils.py`
    - Make sure to run the doc example doc test locally as described in https://github.com/huggingface/transformers/tree/master/docs#for-python-files
    - Optionally, change the example docstring to a more sensible example that gives a better suited result
    - Make the test pass
    - Add the file name to https://github.com/huggingface/transformers/blob/master/utils/documentation_tests.txt (making sure the file stays in alphabetical order)
    - Run the doc example test again locally
4. Open the PR and tag me @patrickvonplaten @ydshieh or @patil-suraj (don't forget to run `make fixup` before your final commit) 🎊 
    - Note that some code is copied across our codebase. If you see a line like `# Copied from transformers.models.bert...`, this means that the code is copied from that source, and our scripts will automatically keep that in sync. If you see that, you should not edit the copied method! Instead, edit the original method it's copied from, and run make fixup to synchronize that across all the copies. Be sure you installed the development dependencies with `pip install -e ".[dev]"`, as described in the contributor guidelines above, to ensure that the code quality tools in `make fixup` can run.

### PyTorch Model Examples added to tests:
- [ ] ALBERT
- [x] BART
- [x] BEiT
- [ ] BERT
- [ ] Bert
- [ ] BigBird
- [x] BigBirdPegasus
- [x] Blenderbot
- [x] BlenderbotSmall
- [ ] CamemBERT
- [ ] Canine
- [ ] CLIP
- [ ] ConvBERT
- [x] ConvNext
- [ ] CTRL
- [x] Data2VecAudio
- [ ] Data2VecText
- [ ] DeBERTa
- [ ] DeBERTa-v2
- [x] DeiT
- [ ] DETR
- [ ] DistilBERT
- [ ] DPR
- [ ] ELECTRA
- [ ] Encoder
- [ ] FairSeq
- [ ] FlauBERT
- [ ] FNet
- [ ] Funnel
- [ ] GPT
- [ ] GPT-J
- [x] Hubert
- [ ] I-BERT
- [ ] ImageGPT
- [ ] LayoutLM
- [ ] LayoutLMv2
- [ ] LED
- [ ] Longformer
- [ ] LUKE
- [ ] LXMERT
- [ ] M2M100
- [x] Marian
- [ ] MaskFormer
- [x] mBART
- [ ] MegatronBert
- [ ] MobileBERT
- [ ] MPNet
- [ ] mT5
- [ ] Nystromformer
- [ ] OpenAI
- [ ] OpenAI
- [x] Pegasus
- [ ] Perceiver
- [x] PLBart
- [x] PoolFormer
- [ ] ProphetNet
- [ ] QDQBert
- [ ] RAG
- [ ] Realm
- [ ] Reformer
- [x] ResNet
- [ ] RemBERT
- [ ] RetriBERT
- [ ] RoBERTa
- [ ] RoFormer
- [x] SegFormer
- [x] SEW
- [x] SEW-D
- [x] SpeechEncoderDecoder
- [x] Speech2Text
- [x] Speech2Text2
- [ ] Splinter
- [ ] SqueezeBERT
- [x] Swin
- [ ] T5
- [ ] TAPAS
- [ ] Transformer-XL
- [ ] TrOCR
- [x] UniSpeech
- [x] UniSpeechSat
- [x] Van
- [x] ViLT
- [x] VisionEncoderDecoder
- [ ] VisionTextDualEncoder
- [ ] VisualBert
- [x] ViT
- [x] ViTMAE
- [x] Wav2Vec2
- [x] WavLM
- [ ] XGLM
- [ ] XLM
- [ ] XLM-RoBERTa
- [ ] XLM-RoBERTa-XL
- [ ] XLMProphetNet
- [ ] XLNet
- [ ] YOSO

### Tensorflow Model Examples added to tests:
- [ ] ALBERT
- [ ] BART
- [ ] BEiT
- [ ] BERT
- [ ] Bert
- [ ] BigBird
- [ ] BigBirdPegasus
- [ ] Blenderbot
- [ ] BlenderbotSmall
- [ ] CamemBERT
- [ ] Canine
- [ ] CLIP
- [ ] ConvBERT
- [ ] ConvNext
- [ ] CTRL
- [ ] Data2VecAudio
- [ ] Data2VecText
- [ ] DeBERTa
- [ ] DeBERTa-v2
- [ ] DeiT
- [ ] DETR
- [ ] DistilBERT
- [ ] DPR
- [ ] ELECTRA
- [ ] Encoder
- [ ] FairSeq
- [ ] FlauBERT
- [ ] FNet
- [ ] Funnel
- [ ] GPT
- [ ] GPT-J
- [ ] Hubert
- [ ] I-BERT
- [ ] ImageGPT
- [ ] LayoutLM
- [ ] LayoutLMv2
- [ ] LED
- [ ] Longformer
- [ ] LUKE
- [ ] LXMERT
- [ ] M2M100
- [ ] Marian
- [ ] MaskFormer
- [ ] mBART
- [ ] MegatronBert
- [ ] MobileBERT
- [ ] MPNet
- [ ] mT5
- [ ] Nystromformer
- [ ] OpenAI
- [ ] OpenAI
- [ ] Pegasus
- [ ] Perceiver
- [ ] PLBart
- [ ] PoolFormer
- [ ] ProphetNet
- [ ] QDQBert
- [ ] RAG
- [ ] Realm
- [ ] Reformer
- [ ] ResNet
- [ ] RemBERT
- [ ] RetriBERT
- [ ] RoBERTa
- [ ] RoFormer
- [ ] SegFormer
- [ ] SEW
- [ ] SEW-D
- [ ] SpeechEncoderDecoder
- [ ] Speech2Text
- [ ] Speech2Text2
- [ ] Splinter
- [ ] SqueezeBERT
- [ ] Swin
- [ ] T5
- [ ] TAPAS
- [ ] Transformer-XL
- [ ] TrOCR
- [ ] UniSpeech
- [ ] UniSpeechSat
- [ ] Van
- [ ] ViLT
- [ ] VisionEncoderDecoder
- [ ] VisionTextDualEncoder
- [ ] VisualBert
- [ ] ViT
- [ ] ViTMAE
- [ ] Wav2Vec2
- [ ] WavLM
- [ ] XGLM
- [ ] XLM
- [ ] XLM-RoBERTa
- [ ] XLM-RoBERTa-XL
- [ ] XLMProphetNet
- [ ] XLNet
- [ ] YOSO

### PR DESCRIPTION
**Pull request by oneraghavan**
# What does this PR do?

modeling_canine has doc test setup by not included in documentation_tests.txt , this PR adds it

Fixes #16292

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@ArthurZucker 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by oneraghavan** [comment]
@ydshieh  Requesting you to review .

**Comment by ydshieh** [comment]
Sorry, I missed this PR. Look it now.

**Comment by ydshieh** [comment]
@oneraghavan , the doctest would fail for `canine` at this moment. We have to add the expected values (loss or some outputs).

Would you like to follow the changes in PR #16441 for `modeling_longformer.py`, more precisely in `LongformerForSequenceClassification` and `LongformerForTokenClassification`. See those changes [here](https://github.com/huggingface/transformers/pull/16441/files).

Don't hesitate if you have any question. Thank you!

**Comment by oneraghavan** [comment]
@ydshieh Will add those changes.

**Comment by ydshieh** [comment]
@oneraghavan Thank you 🤗 . I reopened the PR. Before continue the work, don't forget to update your local `main` branch first, then rebase your working branch on `main` branch.

**Comment by oneraghavan** [comment]
@ydshieh  I request you to reopen the PR again. I have fixed the checkpoints, the tests should pass now.

**Comment by oneraghavan** [comment]
@ydshieh  I think this is good to merge.

File: src/transformers/models/canine/modeling_canine.py [hunk 1]
```diff
@@ -1467,9 +1469,64 @@ def __init__(self, config):
     @add_start_docstrings_to_model_forward(CANINE_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
     @add_code_sample_docstrings(
         processor_class=_TOKENIZER_FOR_DOC,
-        checkpoint=_CHECKPOINT_FOR_DOC,
+        checkpoint="aliosm/sha3bor-poetry-diacritizer-canine-s",
```
Comment (ydshieh) [inline src/transformers/models/canine/modeling_canine.py:1472]: This checkpoint is in Arabic, and trained for diacritizer.

File: src/transformers/models/canine/modeling_canine.py [hunk 2]
```diff
@@ -1467,9 +1469,64 @@ def __init__(self, config):
...
+            "LABEL_3",
+            "LABEL_0",
+            "LABEL_0",
+            "LABEL_0",
+            "LABEL_1",
+            "LABEL_0",
+            "LABEL_3",
+            "LABEL_3",
+        ],
+        expected_loss=0.46,
```
Comment (ydshieh) [inline src/transformers/models/canine/modeling_canine.py:1529]: This list is too long and doesn't really provide meaningful information, as the model is character-based. If the list is short, we could accept it (as done in a few other doc examples).

**Review by ydshieh — COMMENTED** [review]
@oneraghavan The doctest for this model indeed passes!

However, I have some concern for `CanineForTokenClassification`. 

- The checkpoint is for Arabic text, and trained for diacritizer.
- The model is character-based, and the expected output contains a **very** long list of `LABEL_X`.

I would suggest not to use `add_code_sample_docstrings` for `CanineForTokenClassification`, but to overwrite the code sample in the modeling file without checking expected outputs.

@sgugger Any opinion for this?

**Comment by sgugger** [comment]
Yes, agreed @ydshieh . In general any result that is LABEL_0 or a list of those should really not be included.

**Comment by ydshieh** [comment]
I don't think we have an easy way to ignore the doctest in this case. The `>>> predicted_tokens_classes` part in `PT_TOKEN_CLASSIFICATION_SAMPLE` in the file `src/transformers/utils/doc.py` requires some expected outputs for `predicted_tokens_classes`. If there is none, the test just fails.


```python
    >>> predicted_tokens_classes
    {expected_output}
```

**Comment by oneraghavan** [comment]
@ydshieh  @sgugger  Can we do add a paramerter in add_code_sample_docstrings in function and leave the default to None. Then when places we need to use custom sample, we can call it from there .

The function definition will look like this 

def add_code_sample_docstrings(
    *docstr,
    processor_class=None,
    checkpoint=None,
    output_type=None,
    config_class=None,
    mask="[MASK]",
    qa_target_start_index=14,
    qa_target_end_index=15,
    model_cls=None,
    modality=None,
    expected_output="",
    expected_loss="",
    code_sample="",
):

Inside I can use code_sample if it has been passed or look up the code sample from the templates. 

Let me know if this is okay.

**Comment by ydshieh** [comment]
I don't see how [the latest change](https://github.com/huggingface/transformers/pull/18225/commits/e1e98b9576c6a331f2b74f730ddd08c6f47421d6) is better than just putting the docstring under `CanineForTokenClassification` directly.

I will leave @sgugger to give his opinion.

**Comment by sgugger** [comment]
We don't need any other tooling here. Either the model falls in the "automatic docstring" category or it does not. If it does not, we just write the docstring (with the replace return decorator).
